luigi/contrib/hdfs/hadoopcli_clients.py

```diff
@@ -0,0 +1,288 @@
+    """
+    version = hdfs_config.get_configured_hadoop_version()
+    if version == "cdh4":
+        return HdfsClient()
+    elif version == "cdh3":
+        return HdfsClientCdh3()
+    elif version == "apache1":
+        return HdfsClientApache1()
+    else:
+        raise Exception("Error: Unknown version specified in Hadoop version"
+                        "configuration parameter")
+
+
+class HdfsClient(FileSystem):
+    """
+    This client uses Apache 2.x syntax for file system commands, which also matched CDH4.
+    """
+
+    recursive_listdir_cmd = ['-ls', '-R']
+
+    @staticmethod
+    def call_check(command):
+        p = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True, universal_newlines=True)
+        stdout, stderr = p.communicate()
+        if p.returncode != 0:
+            raise hdfs_error.HDFSCliError(command, p.returncode, stdout, stderr)
+        return stdout
+
+    def exists(self, path):
+        """
+        Use ``hadoop fs -stat`` to check file existence.
+        """
+
+        cmd = load_hadoop_cmd() + ['fs', '-stat', path]
+        logger.debug('Running file existence check: %s', u' '.join(cmd))
+        p = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, close_fds=True, universal_newlines=True)
+        stdout, stderr = p.communicate()
+        if p.returncode == 0:
+            return True
+        else:
+            not_found_pattern = "^.*No such file or directory$"
+            not_found_re = re.compile(not_found_pattern)
+            for line in stderr.split('\n'):
+                if not_found_re.match(line):
+                    return False
+            raise hdfs_error.HDFSCliError(cmd, p.returncode, stdout, stderr)
+
+    def rename(self, path, dest):
+        parent_dir = os.path.dirname(dest)
+        if parent_dir != '' and not self.exists(parent_dir):
+            self.mkdir(parent_dir)
+        if type(path) not in (list, tuple):
+            path = [path]
+        else:
+            warnings.warn("Renaming multiple files at once is not atomic.")
+        self.call_check(load_hadoop_cmd() + ['fs', '-mv'] + path + [dest])
+
+    def rename_dont_move(self, path, dest):
+        """
+        Override this method with an implementation that uses rename2,
+        which is a rename operation that never moves.
+
+        For instance, `rename2 a b` never moves `a` into `b` folder.
+
+        Currently, the hadoop cli does not support this operation.
+
+        We keep the interface simple by just aliasing this to
+        normal rename and let individual implementations redefine the method.
+
+        rename2 -
+        https://github.com/apache/hadoop/blob/ae91b13/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/protocol/ClientProtocol.java
+        (lines 483-523)
+        """
+        warnings.warn("Configured HDFS client doesn't support rename_dont_move, using normal mv operation instead.")
```

Comment:
Most of the time warnings produce too much information, and usually these information are misleading or unnecessary(e.g. line info, file name).

> Also, why log.warning? :)

Oh, I was wrong :) I should have read the code more carefully. In this case, logger.info or logger.debug would be more useful since the warning doesn't have any effect in program's workflow. It's just a message that can be useful for debugging(or for similar reasons). Otherwise, it would be better to raise an exception.